### PR TITLE
[BEAM-8383] Add metrics to the Python state cache

### PR DESCRIPTION
**Pull request by mxm**
For more insight into how effectively the cache works, metrics have been added
to the Python SDK. All the state operations are counted, as well as metrics like
the current size of the cache, cache hits/misses, and the capacity of the cache.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by mxm** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by mxm** [comment]
Run Java PreCommit

**Comment by mxm** [comment]
I've ran some cluster tests with this. The metrics provide pretty good insight into how well the state caching performance. The hit/miss gauges as well as the current state size/capacity allows users to tune the cache to a decent size. The metrics are available per-bundle and as totals.

This could need a review now :)

**Comment by mxm** [comment]
Just to summarize, these are the metrics provided:

```
# Per-bundle gauges
'get'
'put'
'extend'
'miss'
'hit'
'clear'
'evict'
'size'
'capacity'
# Counters
'get_total'
'put_total'
'extend_total'
'miss_total'
'hit_total'
'clear_total'
'evict_total'
```

File: runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java [hunk 1]
```diff
@@ -306,8 +307,15 @@ public void update(Iterable<MonitoringInfo> monitoringInfos) {
           if (metric.hasCounterData()) {
             CounterData counterData = metric.getCounterData();
             if (counterData.getValueCase() == CounterData.ValueCase.INT64_VALUE) {
-              Counter counter = getCounter(metricName);
-              counter.inc(counterData.getInt64Value());
+              String gaugeURN =
+                  BeamUrns.getUrn(MetricsApi.MonitoringInfoTypeUrns.Enum.LATEST_INT64_TYPE);
```
Comment (tweise) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/MetricsContainerImpl.java]: `gaugeURN` is constant and can be retrieved once.

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py [hunk 2]
```diff
@@ -219,42 +223,72 @@ def test_flattened_side_input(self):
...
+      experiments = options.view_as(DebugOptions).experiments or []
+      experiments.append('state_cache_size=123')
+      options.view_as(DebugOptions).experiments = experiments
       with Pipeline(self.get_runner(), options) as p:
         # pylint: disable=expression-not-assigned
         (p
-         | beam.Create(list(range(n)))
-         | beam.ParDo(DoFn()))
-
+         | "source" >> FlinkStreamingImpulseSource()
```
Comment (tweise) [inline sdks/python/apache_beam/runners/portability/flink_runner_test.py]: Is it necessary to use the (deprecated) source here? Wouldn't `Create` also produce multiple bundles when  `max_bundle_size` is set?

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py [hunk 3]
```diff
@@ -219,42 +223,72 @@ def test_flattened_side_input(self):
...
+                        # only getting initialized after the first bundle.
+                        # Caching depends on the cache token which is lazily
+                        # initialized by the Runner's StateRequestHandlers).
+                        'stateful.beam.metric:statecache:get_total: 100',
+                        'stateful.beam.metric:statecache:miss_total: 10',
+                        'stateful.beam.metric:statecache:hit_total: 90',
+                        'stateful.beam.metric:statecache:put_total: 10',
+                        'stateful.beam.metric:statecache:extend_total: 100',
+                        'stateful.beam.metric:statecache:evict_total: 0',
+                       }
```
Comment (tweise) [inline sdks/python/apache_beam/runners/portability/flink_runner_test.py]: nice!

File: sdks/python/apache_beam/runners/worker/statecache.py [hunk 4]
```diff
@@ -43,19 +131,23 @@ class StateCache(object):
   def __init__(self, max_entries):
     logging.info('Creating state cache with size %s', max_entries)
     self._cache = self.LRUCache(max_entries, (None, None))
-    self._lock = Lock()
+    self._lock = threading.RLock()
+    self._metrics = Metrics()
 
+  @Metrics.counter_hit_miss("get", "hit", "miss")
```
Comment (tweise) [inline sdks/python/apache_beam/runners/worker/statecache.py]: looks like black magic :)

File: runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/GaugeData.java [hunk 5]
```diff
@@ -57,7 +58,7 @@ public GaugeResult extractResult() {
   public static class EmptyGaugeData extends GaugeData {
 
     private static final EmptyGaugeData INSTANCE = new EmptyGaugeData();
-    private static final Instant EPOCH = new Instant(0);
+    private static final Instant EPOCH = new Instant(GlobalWindow.TIMESTAMP_MIN_VALUE);
```
Comment (tweise) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/GaugeData.java]: Why this change?

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py [hunk 6]
```diff
@@ -219,42 +223,72 @@ def test_flattened_side_input(self):
...
-         | beam.ParDo(DoFn()))
-
+         | "source" >> FlinkStreamingImpulseSource()
+         .set_interval_ms(0).set_message_count(110)
+         | "mapper" >> beam.Map(lambda x: (int(x.decode('utf-8')) % 10, 'val'))
+         | "stateful" >> beam.ParDo(DoFn()))
+
+      lines_expected = {'counter: 110',
+                        # Gauges for the last finished bundle
+                        'stateful.beam.metric:statecache:capacity: 123',
```
Comment (tweise) [inline sdks/python/apache_beam/runners/portability/flink_runner_test.py]: Why are the metrics scoped by transform, aren't they by worker? Are the same metrics reported for each transform that executes in a worker?

File: runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/GaugeData.java (hunk 5 above)
Comment (mxm) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/GaugeData.java]: I think this is a bug because the empty gauge will only be overridden with a timestamp > 0, but any time stamp > min timestamp should override it.

File: sdks/python/apache_beam/runners/portability/flink_runner_test.py (hunk 6 above)
Comment (mxm) [reply sdks/python/apache_beam/runners/portability/flink_runner_test.py]: The metrics are reported per transform because we can have multiple stateful transforms inside an SdkHarness. The only metrics which are per SDK Harness are `size` and `capacity` because the state cache is shared across all the transforms. Metrics currently cannot be reported for the worker itself; that's why these two are duplicated across all transforms. Since the user never knows which Sdk Harness runs the transform, I think it makes sense to always include those two.

**Comment by mxm** [comment]
>Since this throws a few things into the per-record path, have you compared performance (CPU, throughput) before/after?

I've compared the CPU usage and didn't see any difference. Of course there might be a slight overhead but I think it is very small compared to all the other instructions which are involved with state. That said, I didn't observe any CPU increase.

File: sdks/python/apache_beam/runners/worker/statecache.py (hunk 4 above)
Comment (mxm) [reply sdks/python/apache_beam/runners/worker/statecache.py]: Decorators are magical :)

File: runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/GaugeData.java (hunk 5 above)
Comment (tweise) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/GaugeData.java]: Better to rename the constant in this case?

Comment (mxm) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/GaugeData.java]: We can rename it in a follow-up if you don't mind.

**Comment by mxm** [comment]
Run Python PreCommit

**Comment by mxm** [comment]
Run Python PreCommit

**Comment by mxm** [comment]
Run Python PreCommit

**Comment by mxm** [comment]
Builds for Python time out, but tests were consistently passing before on Jenkins and locally. There appears to be an issue with the test infrastructure.
